File: KLASA_3/ROZDZIAL_1/LEKCJA_1/C++/ZAD7.cpp
#include <iostream>
#include <stack>

using namespace std;

bool czy_operand(char znak)
{
    return (znak >= '0' && znak <= '9') || (znak >= 'a' && znak <= 'z');
}

bool czy_operator(char znak)
{
    return (
        znak == '+' ||
        znak == '-' ||
        znak == '*' ||
        znak == '/'
    );
}

bool czy_poprawne(string wyrazenie)
{
    if(wyrazenie.size()<3)return false;
    if(!czy_operand(wyrazenie[0]) || !czy_operand(wyrazenie[1]) || !czy_operator(wyrazenie[wyrazenie.size()-1]))return false;

    int check = 0;
    for(char znak: wyrazenie){
        if(czy_operand(znak)){
            check++;
        }else if(czy_operator(znak)){
            check--;
        }
    }
    if(check!=1)return false;
    return true;
}

int main()
{
    string wyrazenie;
    cout << "Podaj wyrazenie zapisane w ONP do zweryfikowania: ";
    cin >> wyrazenie;

    cout << (czy_poprawne(wyrazenie) ? "TAK" : "NIE");

    return 0;
}

File: KLASA_3/ROZDZIAL_1/LEKCJA_1/PYTHON/ZAD7.py
def czy_operand(char):
    return (char >= '0' and char <= '9') or (char >= 'a' and char <= 'z')


def czy_operator(char):
    return char in '+-*/'


def czy_poprawne(wyrazenie):
    if len(wyrazenie)<3:
        return False
    if (not czy_operand(wyrazenie[0])) or (not czy_operand(wyrazenie[1])) or (not czy_operator(wyrazenie[-1])):
        return False

    check=0
    for char in onp:
        if czy_operand(char):
            check+=1
        elif czy_operator(char):
            check-=1
    if check != 1:
        return False

    return True


onp = input('Podaj wyrażenie zapisane w ONP do zweryfikowania: ')
print('TAK' if czy_poprawne(onp) else 'NIE')